group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java

```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+        }));
+        newAssignment.forEach((targetPartition, members) -> members.forEach(member ->
+            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition)));
+
+        return groupAssignment(finalAssignment, groupSpec.memberIds());
+    }
+
+    private GroupAssignment groupAssignment(
+        Map<String, Set<TopicIdPartition>> assignmentByMember,
+        Collection<String> allGroupMembers
+    ) {
+        Map<String, MemberAssignment> members = new HashMap<>();
+        for (Map.Entry<String, Set<TopicIdPartition>> entry : assignmentByMember.entrySet()) {
+            Map<Uuid, Set<Integer>> targetPartitions = new HashMap<>();
+            entry.getValue().forEach(targetPartition -> targetPartitions.computeIfAbsent(targetPartition.topicId(), k -> new HashSet<>()).add(targetPartition.partitionId()));
+            members.put(entry.getKey(), new MemberAssignmentImpl(targetPartitions));
+        }
+        allGroupMembers.forEach(member -> {
+            if (!members.containsKey(member))
+                members.put(member, new MemberAssignmentImpl(new HashMap<>()));
+        });
+
         return new GroupAssignment(members);
     }
 
-    private Map<Uuid, Set<Integer>> computeTargetPartitions(
-        Set<Uuid> subscribeTopicIds,
+    /**
+     * This function updates assignment by hashing the member IDs of the members and maps the partitions assigned to the members based on the hash. This gives approximately even balance.
+     * @param targetPartitions - the subscribed topic partitions.
+     * @param memberIds - the member ids to which the topic partitions need to be assigned.
+     * @param assignment - the existing assignment by topic partition. We need to pass it as a parameter because this
+     *                   function would be called multiple times for heterogeneous assignment.
+     */
+    void memberHashAssignment(
+        List<TopicIdPartition> targetPartitions,
+        Collection<String> memberIds,
+        Map<TopicIdPartition, List<String>> assignment
+    ) {
+        if (!targetPartitions.isEmpty())
+            for (String memberId : memberIds) {
+                int topicPartitionIndex = Math.abs(memberId.hashCode() % targetPartitions.size());
+                TopicIdPartition topicPartition = targetPartitions.get(topicPartitionIndex);
+                assignment.computeIfAbsent(topicPartition, k -> new ArrayList<>()).add(memberId);
+            }
+    }
+
+    /**
+     * This functions assigns topic partitions to members by round-robin approach and updates the existing assignment.
+     * @param memberIds - the member ids to which the topic partitions need to be assigned.
+     * @param targetPartitions - the subscribed topic partitions.
+     * @param assignment - the existing assignment by topic partition.
+     */
+    void roundRobinAssignment(
+        Collection<String> memberIds,
+        List<TopicIdPartition> targetPartitions,
```

Comment:
targetPartitions seems confusing here.
```suggestion
        List<TopicIdPartition> unassignedPartitions/topicIdPartitions,
```